module/AbstractPurchase.java

```diff
@@ -0,0 +1,40 @@
+package module;
+
+public abstract class AbstractPurchase implements Comparable<AbstractPurchase> {
+    protected final Product product;
+    protected final int units;
+
+    public AbstractPurchase(Product product, int units) {
+        this.product = product;
+        this.units = units;
+    }
+
+    protected abstract Euro getFinalCost(Euro baseCost);
+
+    public Euro getCost() {
+        Euro baseCost = product.getPrice().mul(units);
+        Euro finalCost = getFinalCost(baseCost);
+
+        return finalCost.round(RoundMethods.ROUND, 2);
+    }
+
+    @Override
+    public String toString() {
+        return this.getClass().getSimpleName() + ";" + product.toString() + ";" + units + ";" + this.getCost().toString();

```

Comment:
```suggestion
        return this.getClass().getSimpleName() + ";" + product + ";" + units + ";" + getCost();
```
use `toString` implicitly. 
and don't use `this` unless absolutely necessary.